Add the StashStatusPush handler

ahammond opened the pull request:
This adds a StashStatusPush handler that works with Atlassian Stash to update commits with build status.

sa2ajj commented:
pyflakes error:

```
master/buildbot/status/status_stash.py:9: 'Property' imported but unused
```

sa2ajj commented on master/docs/manual/cfg-statustargets.rst:
```diff
@@ -1561,6 +1561,37 @@ In case any of `repoOwner`, `repoName` or `sha` returns `None`, `False` or empty
+StashStatusPush
+~~~~~~~~~~~~~~~
+
+.. @cindex StashStatusPush
```
`@cindex` is a remnant of texinfo format.

Please see other classes for an example of `.. index::`

ahammond replied:
I actually don't know squat about rst, so I just followed the example for the GitHubStatus entry above.

sa2ajj replied:
Yeah, sorry about that: the docs need to be cleaned up :/

ahammond replied:
I'm going to leave that to be consistent with the rest of the file then.

sa2ajj commented:
There are three things I'd like to bring up:

* at least one test case
* update of release notes
* since we have plugins in eight, update of `master/setup.py` to inlcude this new status target there

sa2ajj commented on master/buildbot/status/status_stash.py:
```diff
@@ -0,0 +1,135 @@
+        :return:
+        """
+        StatusReceiverMultiService.__init__(self)
+        self.base_url = 'https://' + host + '/rest/build-status/1.0/commits/'
```
I'd rather see a format string as below.

sa2ajj commented on master/buildbot/status/status_stash.py:
```diff
@@ -0,0 +1,135 @@
+        """
+        StatusReceiverMultiService.__init__(self)
+        self.base_url = 'https://' + host + '/rest/build-status/1.0/commits/'
+        self.auth = b64encode('{0}:{1}'.format(user, password))
```
Master needs to support python 2.6 where `{0}` format string will not work.

sa2ajj commented on master/buildbot/status/status_stash.py:
```diff
@@ -0,0 +1,135 @@
+        return self
+
+    def buildStarted(self, builderName, build):
+        self.send(builderName, build, 'INPROGRESS')
```
I'd suggest to make these special names -- `INPROGRESS`, `SUCCESSFUL`, `FAILED` -- global private constants.

sa2ajj commented on master/docs/manual/cfg-statustargets.rst:
```diff
@@ -1561,6 +1561,37 @@ In case any of `repoOwner`, `repoName` or `sha` returns `None`, `False` or empty
+    c['status'].append(ss)
+
+
+:class:`StashStatusPush` publishes build status using `Stash Build Integration REST API <https://developer.atlassian.com/static/rest/stash/3.6.0/stash-build-integration-rest.html>`_.
```
I'd somehow emphasize that the status is published for each **build**

sa2ajj commented:
Thank you for your PR :)  Please let me know if you need help.

ahammond commented:
@sa2ajj I think that covers all your comments.

sa2ajj commented on master/docs/manual/cfg-statustargets.rst:
```diff
@@ -1542,7 +1542,9 @@ GitHubStatus
 
-:class:`GitHubStatus` publishes a build status using `GitHub Status API <http://developer.github.com/v3/repos/statuses>`_.
+:class:`GitHubStatus` publishes build status using `GitHub Status API <http://developer.github.com/v3/repos/statuses>`_.
+The build status is published to a specific commit SHA in Stash.
```
I think it's a typo here: Stash.

sa2ajj commented:
Yes, it does.  Thank you! :)

There's a typo introduced though.

And I'll wait for Travis to show something green :)

sa2ajj commented:
Thanks.  Now it's just Travis.

sa2ajj commented:
Problem with a test

```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/trial/runner.py", line 561, in loadPackage
    module = modinfo.load()
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/modules.py", line 383, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/_reflectpy3.py", line 268, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/_reflectpy3.py", line 211, in _importAndCheckStack
    reraise(excValue, excTraceback)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_status_stash.py", line 18, in <module>
    from buildbot.status.status_stash import StashStatusPush, INPROGRESS, SUCCESSFUL, FAILED
  File "/home/travis/build/buildbot/buildbot/master/buildbot/status/status_stash.py", line 14, in <module>
    from twisted.web.client import Agent, readBody
exceptions.ImportError: cannot import name readBody

buildbot.test.unit.test_status_stash
```

sa2ajj commented:
It looks like it works with Twisted v12.x, while fails for Twisted 13 & 14

ahammond commented:
Strange. I tested it on Twisted 13.2

On Wed, Jan 28, 2015 at 2:50 PM, Mikhail Sobolev <notifications@github.com>
wrote:

> It looks like it works with Twisted v12.x, while fails for Twisted 13 & 14
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/buildbot/buildbot/pull/1508#issuecomment-71933675>.
>

ahammond commented:
http://twistedmatrix.com/documents/current/api/twisted.web.client.readBody.html

sa2ajj commented:
I said "it looks like" -- based on the matrix Travis shows.

sa2ajj commented:
Interestingly, the test succeed for me in this virtual env:

```
argparse (1.2.1)
buildbot (0.8.11-pre-95-g2dbac6b)
decorator (3.4.0)
Jinja2 (2.7.3)
MarkupSafe (0.23)
mock (1.0.1)
pip (1.5.6)
python-dateutil (2.4.0)
setuptools (5.5.1)
six (1.9.0)
SQLAlchemy (0.7.10)
sqlalchemy-migrate (0.7.2)
Tempita (0.5.2)
Twisted (14.0.2)
wsgiref (0.1.2)
zope.interface (4.1.2)
```

ahammond commented on master/buildbot/status/status_stash.py:
```diff
@@ -0,0 +1,141 @@
+
+    def buildFinished(self, builderName, build, results):
+        self.send(builderName, build,
+                  SUCCESSFUL if results == SUCCESS else FAILED)
```
@sa2ajj this doesn't appear to be working. I have builds that are passing that are being reported as FAILED. So... another issue. I'll look into it now.

ahammond commented:
@sa2ajj ok, I merged it to catch up with eight. UTs pass regularly on the environments I've tried it in (2.6 and 2.7 with twisted 14). I tested it against a trivial repository with a build that always succeeds and it correctly marked it as successful, so I'm guessing the false negative reports are something to do with a big ugly build that I was testing it against more than anything else.

sa2ajj commented:
@ahammond Please do a rebase.  Such merges make it difficult to impossible to port the change should it be relevant.

Please let me know if you need help.

ahammond commented:
Rebased, and squashed to a single commit for pretty history. I also changed the init() to take a base_url rather than host so that people can decide for themselves if they prefer http or https. Updated docs to reflect this change.

sa2ajj commented:
@ahammond thank you.

We still have this strange error :(  I'm trying to figure out what's wrong.

sa2ajj commented:
We have pyflakes errors:

```
pyflakes master/buildbot slave/buildslave
master/buildbot/test/unit/test_status_stash.py:20: 'fakedb' imported but unused
master/buildbot/test/unit/test_status_stash.py:21: 'fakemaster' imported but unused
master/buildbot/test/unit/test_status_stash.py:24: 'defer' imported but unused
```

And @skelly said that `readBody` is only available w/ Twisted v13.1.0+

(I'm checking why builds w/ Twisted 12.0 did not fail)

sa2ajj commented:
I still do not quite understand why the test is not failing in Travis (looks like it's skipped in a fancy way for some reason).

I run the test locally w/ Twisted 12.0.0 and it does indeed fail.

Since the requirement is Twisted >= 11.0.0, this somehow needs to be accomodated (@djmitche, do we have a reason to be still compatible w/ it?)

ahammond commented:
I removed the unused imports from the test code.

sa2ajj commented:
So I made an experiment (GH-1529) and it turned out that installing `txgithub` when Twisted is too old for it (the requirement is '>= 12.3.0), the latter gets upgraded to the latest one and the test passes.

@ahammond the code need to be adapted so it would work with versions of Twisted supported by Buildbot.

sa2ajj commented:
@ahammond alternative is to check gracefully if the function is available and refuse to work if it does not.

djmitche commented:
I think this is a relatively straightforward fix.  The problem with requiring a newer Twisted is that it causes problems for downstream packager where they must wait for those newer Twisted versions to stabilize.  If this turns out to be a difficult fix, we can go down that route, but in general I try to hold out as long as practical before increasing python requirements.

ahammond commented:
I wrapped the readBody() import in a try with a stub function that, in the event of an Stash API failure, will tell the user to upgrade to Twisted>=13.1 if they want verbose error messages. This should solve the import issue.

sa2ajj commented:
Thank you, @ahammond 

There's a little problem with `pep8` being too new.  So I created GH-1537 where the corresponding change from master is cherry-picked.

sa2ajj commented on master/buildbot/status/status_stash.py:
```diff
@@ -0,0 +1,151 @@
+            method='POST',
+            uri=bytes(stash_uri),
+            headers=Headers({
+                'Content-Type': ['application/json; charset=utf-8',],
```
One missing space:

```
pep8 --config=common/pep8rc master slave
master/buildbot/status/status_stash.py:101:67: E231 missing whitespace after ','
                'Content-Type': ['application/json; charset=utf-8',],
                                                                  ^
1       E231 missing whitespace after ','
1
make: *** [pep8] Error 1
```

Do you need a comma here at all?

ahammond commented:
added pep8 silencing space.

sa2ajj commented:
Thank you very much for being patient and addressing review comments.

And thank you for your contribution, @ahammond
